credential_provider/pod_identity_credential_provider.go

```diff
@@ -117,94 +55,55 @@ func (p *podIdentityTokenFetcher) FetchToken(ctx credentials.Context) ([]byte, e
 	if err != nil {
-		return nil, err
+		return "", err
 	}
 
-	return []byte(tokRsp.Status.Token), nil
+	return tokRsp.Status.Token, nil
 }
 
-func (p *PodIdentityCredentialProvider) GetAWSConfig() (*aws.Config, error) {
-	// Get token for Pod Identity
-	token, tokenErr := p.fetcher.FetchToken(context.Background())
-	if tokenErr != nil {
-		return nil, fmt.Errorf("failed to fetch token: %+v", tokenErr)
-	}
-
-	var config *aws.Config
-	var configErr error
-	if p.preferredEndpoint == preferenceIPv4 || p.preferredEndpoint == preferenceAuto {
-		config, configErr = p.getAWSConfigFromPodIdentityAgent(token, podIdentityAgentEndpointIPv4)
-		if configErr != nil {
-			klog.Warningf("IPv4 endpoint attempt failed: %+v.", configErr)
-		} else {
-			return config, nil
-		}
-	}
-
-	if p.preferredEndpoint == preferenceIPv6 || p.preferredEndpoint == preferenceAuto {
-		config, configErr = p.getAWSConfigFromPodIdentityAgent(token, podIdentityAgentEndpointIPv6)
-		if configErr != nil {
-			klog.Warningf("IPv6 endpoint attempt failed: %+v.", configErr)
-		}
-	}
-
-	if configErr != nil {
-		return nil, fmt.Errorf("failed to get AWS config from pod identity agent: %+v", configErr)
-	}
-
-	return config, nil
+// PodIdentityCredentialProvider implements CredentialProvider using pod identity
+type PodIdentityCredentialProvider struct {
+	region             string
+	credentialEndpoint string
+	fetcher            endpointcreds.AuthTokenProvider
+	httpClient         *http.Client
 }
 
-func (p *PodIdentityCredentialProvider) getAWSConfigFromPodIdentityAgent(token []byte, podIdentityAgentEndpoint string) (*aws.Config, error) {
-	req, err := http.NewRequest("GET", podIdentityAgentEndpoint, nil)
-	if err != nil {
-
-		return nil, fmt.Errorf("failed to create HTTP request to pod identity agent: %+v", err)
-	}
-	req.Header.Set(podIdentityAuthHeader, string(token))
-	resp, err := p.httpClient.Do(req)
-	if err != nil {
-		return nil, fmt.Errorf("HTTP request to pod identity agent failed: %+v", err)
-	}
-
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		// Read the response body
-		body, err := io.ReadAll(resp.Body)
-		if err != nil {
-			return nil, fmt.Errorf("failed to read error response body: %v, status code: %d", err, resp.StatusCode)
-		}
-
-		return nil, fmt.Errorf("pod identity agent returned error - Status: %d, Headers: %v, Body: %s",
-			resp.StatusCode,
-			resp.Header,
-			string(body))
-	}
-
-	var creds struct {
-		AccessKeyId     string `json:"AccessKeyId"`
-		SecretAccessKey string `json:"SecretAccessKey"`
-		Token           string `json:"Token"`
-	}
+func NewPodIdentityCredentialProvider(
+	region, nameSpace, svcAcc, podName, preferredAddressType string,
+	k8sClient k8sv1.CoreV1Interface,
+) (ConfigProvider, error) {
 
-	if err := json.NewDecoder(resp.Body).Decode(&creds); err != nil {
-		return nil, fmt.Errorf("failed to decode credentials from pod identity agent: %+v", err)
+	endpoint := podIdentityAgentEndpointIPv4
```

Comment:
As commented earlier, this just defaults to IPv4 unless IPv6 is explicitly specified. Validation is pushed into the Mount handler